src/subdataframe/subdataframe.jl

```diff
@@ -176,32 +176,106 @@ Base.@propagate_inbounds Base.getindex(df::SubDataFrame, row_ind::typeof(!),
                                        col_inds::MultiColumnIndex) =
     select(df, col_inds, copycols=false)
 
-
 Base.@propagate_inbounds function Base.setindex!(sdf::SubDataFrame, val::Any, idx::CartesianIndex{2})
-    setindex!(sdf, val, idx[1], idx[2])
+    return setindex!(sdf, val, idx[1], idx[2])
 end
+
 Base.@propagate_inbounds function Base.setindex!(sdf::SubDataFrame, val::Any, ::Colon, colinds::Any)
-    parent(sdf)[rows(sdf), parentcols(index(sdf), colinds)] = val
+    if colinds isa SymbolOrString && columnindex(sdf, colinds) == 0
+        if !is_column_adding_allowed(sdf)
+            throw(ArgumentError("creating new columns in a SubDataFrame that subsets " *
+                                "columns of its parent data frame is disallowed"))
+        end
+        if !(val isa AbstractVector && nrow(sdf) == length(val))
+            throw(ArgumentError("Assigned value must be a vector with length " *
+                                "equal to number of rows in the SubDataFrame"))
+        end
+        T = eltype(val)
+        newcol = Tables.allocatecolumn(Union{T, Missing}, nrow(parent(sdf)))
+        fill!(newcol, missing)
+        newcol[rows(sdf)] = val
+        parent(sdf)[!, colinds] = newcol
+    else
+        parent(sdf)[rows(sdf), parentcols(index(sdf), colinds)] = val
+    end
+    return sdf
+end
+
+function Base.setindex!(sdf::SubDataFrame, v::AbstractVector,
+                        ::typeof(!), col_ind::ColumnIndex)
+    if col_ind isa Union{Signed, Unsigned} && !(1 <= col_ind <= ncol(sdf))
+        throw(ArgumentError("Cannot assign to non-existent column: $col_ind"))
+    end
+    if col_ind isa SymbolOrString && columnindex(sdf, col_ind) == 0
+        if !is_column_adding_allowed(sdf)
+            throw(ArgumentError("creating new columns in a SubDataFrame that subsets " *
+                                "columns of its parent data frame is disallowed"))
+        end
+        sdf[:, col_ind] = v
```

Comment:
Should we disallow either `sdf[:, col_ind]` or `sdf[!, col_ind]`? Given that the implementations are so similar and the logic about memory access doesn't apply as closely?

Reply:
Both have their uses. The only case in which they are identical is when creating a new column, but this is the same as for `DataFrame`.

Reply:
Exactly. See my comment above. It was for `cols` selector, but it also applies to `col_ind` selector. Sometimes you want in-place operation with conversion (then do `:`) and sometimes you want new allocation with promotion (then do `!`).

Reply:
> Both have their uses. The only case in which they are identical is when creating a new column, but this is the same as for `DataFrame`.

Wait this isn't true. `df[:, :x] = y` and `df[!, :x] = y` mean different things. The first copies `y` and the second doesn't. Usually this is more performant, but in the case of `sdf[:, :x] = y`, using `!` is actually *less* performant. I'm worried this will trip up new users.

Reply:
The mental model should be the following:
- `df[:, :x] = y` is an in-place operation; and this is true with this PR both for `DataFrame` and `SubDataFrame` (except when `:x` is not present in `df`, but I guess you are not worried about this case)
- `df[:, :x] .= y` follows the same rules
- `df[!, :x] = y` is a column replacement operation. A new column `:x` is stored in `df`; and again with this PR this is true both for `DataFrame` and for `SubDataFrame`; the difference is that for `DataFrame` we do not copy `y`, but store it "as is", while for `SubDataFrame` we allocate a new column (it is impossible not to allocate)
- `df.x = y` the same as above
- `df[!, :x] .= y` is a column replacement operation with guaranteed copying of a column. A new column `:x` is stored in `df`; and again with this PR this is true both for `DataFrame` and for `SubDataFrame`; and here in both cases new column is allocated and data is copied
- `df.x .= y` will be the same as above when Julia 1.7 is out

This is the way we should explain the difference between `:`, `!`, and `getproperty`. The design is not about performance, but about allowing different kind of operations. The performance is just a consequence.

In summary - I think new users should not be taught what is fast and what is slow, but rather what is the objective of the operation (and in https://github.com/JuliaData/DataFrames.jl/pull/2794#discussion_r693241259 I have shown an exact case when in-place and replace differ; this is also the same for `DataFrame`).

Reply:
Okay. Thanks for the summary / reminder. 

I'm still on the fence about allowing `df[!, :x]`. I think the scenario where someone would want to avoid type promotion, as in your `Int` and `Char`  example, is relatively uncommon. Comparead to people thinking `!` means speed. 

But the proposed design is internally consistent, so overall I think it's okay to move ahead.

Reply:
> is relatively uncommon

I think the crucial common case is if you want to introduce `missing` in a column that does not allow it.

---

Still, before tackling @nalimilan comments I think it is important to work out this case. I would prefer not to add `!` (and in consequence also we will not add `sdf.newcol = v`) if we feel it is not needed.

I designed a full proposal here, as my experience in the past was that when I proposed a limited functionality I was immediately asked to propose a full one 😄.

@nalimilan - can you please comment what you think here? Thank you!

Reply:
Also @matthieugomez - maybe you have an opinion here?

Reply:
For more background, a common Stata idiom is to go 

```
foreach i in numlist 1/100 {
    replace x = y_`i' if x == `i'
}
```

Which I would imagine a lot of users would be doing here.

Reply:
@pdeffebach the discussion prompted me with two things:
* maybe `sdf[!, :col] = v` if `eltype(v) <: eltype(sdf.col)` should not allocate the column but be in-place if `:col` is present in `sdf`? This would improve efficiency at the cost of docstring complexity. I think it is OK not co copy if not needed as we assume `parent(sdf)` owns `:col`. @nalimilan - what do you think?
* your example currently would be written (if I read it correctly):
```
for i in 1:100
    transform!(view(df, df.x .== i, :), Symbol("y_", i) => :x)
    # or
    transform!(filter(:x => ==(i), df, view=true), Symbol("y_", i) => :x)
    # or
    sdf = view(df, df.x .== i, :x)
    sdf.x = sdf[!, Symbol("y_", i)]
end
```
but probably you meant to make it even more convenient as indeed it is more verbose than Stata (do you have any specific syntax in mind)?

Reply:
I would write this specific example as `df.x = string.("y_", df.x)`. But probably this illustrates a more general idiom? :-D

Anyway, I feel like we don't really have the choice if we want to remain consistent. The rules are already quite complex. Making `df[!, col] .= x` replace columns but `sdf[!, col] .= x` fill in-place (always or only in some particular cases) would be really hard to explain.

Reply:
I think I'm in favor of auto-promoting only when needed. though you are right that can lead to some difficult-to-explain rules and perhaps some performance problems.

Reply:
> I think I'm in favor of auto-promoting only when needed

This is the rule that is currently implemented (we do only promote when needed).

AFAICT what @nalimilan writes is that `sdf[!, col] = x` should always replace `col` in `parent(sdf)` no matter if promotion is needed or not.

What we need to do is:
1. clearly explain in the docs the consequences of `!` on LHS in `sdf`.
2. change the implementation of `select!` and `transform!` so that they do not copy if not needed (I will change this in this PR)